File: tally-android/app/src/main/kotlin/app/tally/observability/TallyLogger.kt
package app.tally.observability

import android.util.Log
import io.sentry.Breadcrumb
import io.sentry.Sentry
import io.sentry.SentryLevel
import org.json.JSONObject
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock
import app.tally.BuildConfig

/**
 * Log level for structured logging
 */
enum class LogLevel(val priority: Int, val value: String) {
    DEBUG(0, "debug"),
    INFO(1, "info"),
    WARN(2, "warn"),
    ERROR(3, "error")
}

/**
 * Context for structured log entries
 */
data class LogContext(
    // Correlation
    val traceId: String? = null,
    val spanId: String? = null,
    val requestId: String? = null,

    // User context (privacy: use hashed IDs only)
    val userId: String? = null,

    // Operation context
    val operation: String? = null,
    val durationMs: Long? = null,

    // Error context
    val error: Throwable? = null,

    // Custom attributes
    val extras: Map<String, Any> = emptyMap()
) {
    /**
     * Create a copy with additional extras
     */
    fun withExtras(vararg pairs: Pair<String, Any>): LogContext {
        return copy(extras = extras + pairs.toMap())
    }
}

/**
 * Structured logger for Android with Sentry integration
 *
 * Provides wide-event / canonical log lines that include correlation IDs,
 * structured context, and integrate with Sentry for error tracking.
 *
 * Usage:
 * ```kotlin
 * import app.tally.observability.TallyLogger
 * import app.tally.observability.LogContext
 *
 * // Simple log
 * TallyLogger.info("User signed in", LogContext(userId = "u_abc123"))
 *
 * // Wide event (canonical log line)
 * TallyLogger.info("api.request.completed", LogContext(
 *     operation = "createChallenge",
 *     durationMs = 150L,
 *     userId = "u_abc123",
 *     extras = mapOf("statusCode" to 200)
 * ))
 *
 * // Error with context
 * TallyLogger.error("Failed to create entry", LogContext(
 *     error = ex,
 *     userId = "u_abc123",
 *     extras = mapOf("challengeId" to "ch_xyz")
 * ))
 * ```
 */
object TallyLogger {
    private const val TAG = "Tally"
    private const val SERVICE = "tally-android"
    private const val PLATFORM = "android"

    private val lock = ReentrantLock()

    @Volatile
    private var minLevel: LogLevel = if (BuildConfig.DEBUG) LogLevel.DEBUG else LogLevel.INFO

    private val isoDateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US).apply {
        timeZone = TimeZone.getTimeZone("UTC")
    }

    /**
     * Set minimum log level
     */
    fun setMinLevel(level: LogLevel) {
        lock.withLock { minLevel = level }
    }

    /**
     * Create a child logger with preset context
     */
    fun child(context: LogContext): ChildLogger {
        return ChildLogger(context)
    }

    // MARK: - Log Methods

    fun debug(message: String, context: LogContext = LogContext()) {
        log(LogLevel.DEBUG, message, context)
    }

    fun info(message: String, context: LogContext = LogContext()) {
        log(LogLevel.INFO, message, context)
    }

    fun warn(message: String, context: LogContext = LogContext()) {
        log(LogLevel.WARN, message, context)
    }

    fun error(message: String, context: LogContext = LogContext()) {
        log(LogLevel.ERROR, message, context)
    }

    // MARK: - Core Logging

    private fun log(level: LogLevel, message: String, context: LogContext) {
        val currentMinLevel = lock.withLock { minLevel }
        if (level.priority < currentMinLevel.priority) return

        val entry = buildLogEntry(level, message, context)

        // Output to Android Log
        when (level) {
            LogLevel.DEBUG -> Log.d(TAG, entry)
            LogLevel.INFO -> Log.i(TAG, entry)
            LogLevel.WARN -> Log.w(TAG, entry)
            LogLevel.ERROR -> Log.e(TAG, entry)
        }

        // Add as Sentry breadcrumb
        val breadcrumb = Breadcrumb().apply {
            this.message = message
            this.category = "log"
            this.level = sentryLevel(level)
            contextToMap(context).forEach { (key, value) ->
                setData(key, value.toString())
            }
        }
        Sentry.addBreadcrumb(breadcrumb)

        // Report errors to Sentry
        if (level == LogLevel.ERROR) {
            context.error?.let { error ->
                Sentry.configureScope { scope ->
                    contextToMap(context).forEach { (key, value) ->
                        scope.setExtra(key, value.toString())
                    }
                }
                Sentry.captureException(error)
            } ?: run {
                Sentry.captureMessage(message)
            }
        }
    }

    private fun buildLogEntry(level: LogLevel, message: String, context: LogContext): String {
        val json = JSONObject().apply {
            put("timestamp", isoDateFormat.format(Date()))
            put("level", level.value)
            put("message", message)
            put("service", SERVICE)
            put("platform", PLATFORM)
            put("environment", if (BuildConfig.DEBUG) "development" else "production")
            put("version", "${BuildConfig.VERSION_NAME}+${BuildConfig.VERSION_CODE}")

            // Add context
            val contextJson = JSONObject()
            contextToMap(context).forEach { (key, value) ->
                contextJson.put(key, value)
            }
            put("context", contextJson)
        }

        return json.toString()
    }

    private fun contextToMap(context: LogContext): Map<String, Any> {
        val map = context.extras.toMutableMap()

        context.traceId?.let { map["traceId"] = it }
        context.spanId?.let { map["spanId"] = it }
        context.requestId?.let { map["requestId"] = it }
        context.userId?.let { map["userId"] = it }
        context.operation?.let { map["operation"] = it }
        context.durationMs?.let { map["duration_ms"] = it }
        context.error?.let { map["errorMessage"] = it.message ?: it.toString() }

        // Remove PII fields
        val piiFields = listOf("email", "password", "token", "secret", "apiKey")
        piiFields.forEach { map.remove(it) }

        return map
    }

    private fun sentryLevel(level: LogLevel): SentryLevel {
        return when (level) {
            LogLevel.DEBUG -> SentryLevel.DEBUG
            LogLevel.INFO -> SentryLevel.INFO
            LogLevel.WARN -> SentryLevel.WARNING
            LogLevel.ERROR -> SentryLevel.ERROR
        }
    }
}

/**
 * Child logger with preset context
 */
class ChildLogger(private val baseContext: LogContext) {

    fun debug(message: String, context: LogContext = LogContext()) {
        TallyLogger.debug(message, mergeContext(context))
    }

    fun info(message: String, context: LogContext = LogContext()) {
        TallyLogger.info(message, mergeContext(context))
    }

    fun warn(message: String, context: LogContext = LogContext()) {
        TallyLogger.warn(message, mergeContext(context))
    }

    fun error(message: String, context: LogContext = LogContext()) {
        TallyLogger.error(message, mergeContext(context))
    }

    private fun mergeContext(context: LogContext): LogContext {
        return LogContext(
            traceId = context.traceId ?: baseContext.traceId,
            spanId = context.spanId ?: baseContext.spanId,
            requestId = context.requestId ?: baseContext.requestId,
            userId = context.userId ?: baseContext.userId,
            operation = context.operation ?: baseContext.operation,
            durationMs = context.durationMs ?: baseContext.durationMs,
            error = context.error ?: baseContext.error,
            extras = baseContext.extras + context.extras
        )
    }
}

File: tally-ios/TallyCore/Sources/TallyCore/Observability/TallyLogger.swift
import Foundation
import Sentry
import os.log

/// Log level for structured logging
public enum LogLevel: String, Comparable {
  case debug = "debug"
  case info = "info"
  case warn = "warn"
  case error = "error"
  
  private var priority: Int {
    switch self {
    case .debug: return 0
    case .info: return 1
    case .warn: return 2
    case .error: return 3
    }
  }
  
  public static func < (lhs: LogLevel, rhs: LogLevel) -> Bool {
    lhs.priority < rhs.priority
  }
}

/// Context for structured log entries
public struct LogContext {
  // Correlation
  public var traceId: String?
  public var spanId: String?
  public var requestId: String?
  
  // User context (privacy: use hashed IDs only)
  public var userId: String?
  
  // Operation context
  public var operation: String?
  public var durationMs: Int?
  
  // Error context
  public var error: Error?
  
  // Custom attributes
  public var extras: [String: Any]
  
  public init(
    traceId: String? = nil,
    spanId: String? = nil,
    requestId: String? = nil,
    userId: String? = nil,
    operation: String? = nil,
    durationMs: Int? = nil,
    error: Error? = nil,
    extras: [String: Any] = [:]
  ) {
    self.traceId = traceId
    self.spanId = spanId
    self.requestId = requestId
    self.userId = userId
    self.operation = operation
    self.durationMs = durationMs
    self.error = error
    self.extras = extras
  }
}

/// Structured logger for iOS with Sentry integration
///
/// Provides wide-event / canonical log lines that include correlation IDs,
/// structured context, and integrate with Sentry for error tracking.
///
/// Usage:
/// ```swift
/// import TallyCore
///
/// // Simple log
/// TallyLogger.shared.info("User signed in", context: LogContext(userId: "u_abc123"))
///
/// // Wide event (canonical log line)
/// TallyLogger.shared.info("api.request.completed", context: LogContext(
///   operation: "createChallenge",
///   durationMs: 150,
///   userId: "u_abc123",
///   extras: ["statusCode": 200]
/// ))
///
/// // Error with context
/// TallyLogger.shared.error("Failed to create entry", context: LogContext(
///   error: err,
///   userId: "u_abc123",
///   extras: ["challengeId": "ch_xyz"]
/// ))
/// ```
public final class TallyLogger: @unchecked Sendable {
  public static let shared = TallyLogger()
  
  private let osLog = OSLog(subsystem: "app.tally", category: "general")
  private let lock = NSLock()
  
  private var _minLevel: LogLevel = .info
  private var minLevel: LogLevel {
    get { lock.withLock { _minLevel } }
    set { lock.withLock { _minLevel = newValue } }
  }
  
  private let service = "tally-ios"
  private let platform = "ios"
  
  private init() {
    #if DEBUG
    _minLevel = .debug
    #endif
  }
  
  /// Set minimum log level
  public func setMinLevel(_ level: LogLevel) {
    minLevel = level
  }
  
  /// Create a child logger with preset context
  public func child(context: LogContext) -> ChildLogger {
    ChildLogger(parent: self, baseContext: context)
  }
  
  // MARK: - Log Methods
  
  public func debug(_ message: String, context: LogContext = LogContext()) {
    log(.debug, message: message, context: context)
  }
  
  public func info(_ message: String, context: LogContext = LogContext()) {
    log(.info, message: message, context: context)
  }
  
  public func warn(_ message: String, context: LogContext = LogContext()) {
    log(.warn, message: message, context: context)
  }
  
  public func error(_ message: String, context: LogContext = LogContext()) {
    log(.error, message: message, context: context)
  }
  
  // MARK: - Core Logging
  
  private func log(_ level: LogLevel, message: String, context: LogContext) {
    guard level >= minLevel else { return }
    
    let entry = buildLogEntry(level: level, message: message, context: context)
    
    // Output to os.log
    let osLogType: OSLogType
    switch level {
    case .debug: osLogType = .debug
    case .info: osLogType = .info
    case .warn: osLogType = .default
    case .error: osLogType = .error
    }
    
    os_log("%{public}@", log: osLog, type: osLogType, entry)
    
    // Add as Sentry breadcrumb
    let breadcrumb = Breadcrumb()
    breadcrumb.message = message
    breadcrumb.category = "log"
    breadcrumb.level = sentryLevel(from: level)
    breadcrumb.data = contextToDict(context)
    SentrySDK.addBreadcrumb(breadcrumb)
    
    // Report errors to Sentry
    if level == .error {
      if let error = context.error {
        SentrySDK.configureScope { scope in
          for (key, value) in self.contextToDict(context) {
            scope.setExtra(value: value, key: key)
          }
        }
        SentrySDK.capture(error: error)
      } else {
        SentrySDK.capture(message: message)
      }
    }
  }
  
  private func buildLogEntry(level: LogLevel, message: String, context: LogContext) -> String {
    var dict: [String: Any] = [
      "timestamp": ISO8601DateFormatter().string(from: Date()),
      "level": level.rawValue,
      "message": message,
      "service": service,
      "platform": platform,
      "environment": environment(),
      "version": appVersion()
    ]
    
    // Add context
    dict["context"] = contextToDict(context)
    
    // Convert to JSON
    if let jsonData = try? JSONSerialization.data(withJSONObject: dict, options: []),
       let jsonString = String(data: jsonData, encoding: .utf8) {
      return jsonString
    }
    
    return "[\(level.rawValue)] \(message)"
  }
  
  private func contextToDict(_ context: LogContext) -> [String: Any] {
    var dict: [String: Any] = context.extras
    
    if let traceId = context.traceId { dict["traceId"] = traceId }
    if let spanId = context.spanId { dict["spanId"] = spanId }
    if let requestId = context.requestId { dict["requestId"] = requestId }
    if let userId = context.userId { dict["userId"] = userId }
    if let operation = context.operation { dict["operation"] = operation }
    if let durationMs = context.durationMs { dict["duration_ms"] = durationMs }
    if let error = context.error { dict["errorMessage"] = error.localizedDescription }
    
    // Remove PII fields
    let piiFields = ["email", "password", "token", "secret", "apiKey"]
    for field in piiFields {
      dict.removeValue(forKey: field)
    }
    
    return dict
  }
  
  private func sentryLevel(from level: LogLevel) -> SentryLevel {
    switch level {
    case .debug: return .debug
    case .info: return .info
    case .warn: return .warning
    case .error: return .error
    }
  }
  
  private func environment() -> String {
    #if DEBUG
    return "development"
    #else
    return "production"
    #endif
  }
  
  private func appVersion() -> String {
    guard let version = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String,
          let build = Bundle.main.infoDictionary?["CFBundleVersion"] as? String else {
      return "unknown"
    }
    return "\(version)+\(build)"
  }
}

/// Child logger with preset context
public final class ChildLogger: @unchecked Sendable {
  private let parent: TallyLogger
  private let baseContext: LogContext
  
  init(parent: TallyLogger, baseContext: LogContext) {
    self.parent = parent
    self.baseContext = baseContext
  }
  
  public func debug(_ message: String, context: LogContext = LogContext()) {
    parent.debug(message, context: mergeContext(context))
  }
  
  public func info(_ message: String, context: LogContext = LogContext()) {
    parent.info(message, context: mergeContext(context))
  }
  
  public func warn(_ message: String, context: LogContext = LogContext()) {
    parent.warn(message, context: mergeContext(context))
  }
  
  public func error(_ message: String, context: LogContext = LogContext()) {
    parent.error(message, context: mergeContext(context))
  }
  
  private func mergeContext(_ context: LogContext) -> LogContext {
    var merged = baseContext
    if let v = context.traceId { merged.traceId = v }
    if let v = context.spanId { merged.spanId = v }
    if let v = context.requestId { merged.requestId = v }
    if let v = context.userId { merged.userId = v }
    if let v = context.operation { merged.operation = v }
    if let v = context.durationMs { merged.durationMs = v }
    if let v = context.error { merged.error = v }
    for (key, value) in context.extras {
      merged.extras[key] = value
    }
    return merged
  }
}
